lang/java/avro/src/main/java/org/apache/avro/generic/GenericDatumWriter.java

```diff
@@ -174,13 +174,22 @@ protected void writeWithoutConversion(Schema schema, Object datum, Encoder out)
         error(schema, datum);
       }
     } catch (NullPointerException e) {
-      throw npe(e, " of " + schema.getFullName());
+      throw npeCceAte(e, " of " + schema.getFullName());
     }
   }
 
-  /** Helper method for adding a message to an NPE. */
-  protected NullPointerException npe(NullPointerException e, String s) {
-    NullPointerException result = new NullPointerException(e.getMessage() + s);
+  /** Helper method for adding a message to an NPE CCE ATE. */
+  protected <T extends Exception> T npeCceAte(T e, String s) {
+    T result = null;
+    if (e instanceof NullPointerException) {
+      result = (T) new NullPointerException(e.getMessage() + s);
+    } else if (e instanceof ClassCastException) {
+      result = (T) new ClassCastException(e.getMessage() + s);
+    } else if (e instanceof AvroTypeException) {
+      result = (T) new AvroTypeException(e.getMessage() + s);
+    } else {
+      result = (T) new Exception(e.getMessage() + s);
```

Comment:
I think this would always throw a ClassCastException itself, If we ever reach this code and T is any subclass of Exception!